File: cpp/sorting/counting_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

vector<int> countingSort(const vector<int> &arr){
    if(arr.empty()) return {};
    int max_val = *max_element(arr.begin(), arr.end());
    vector<int> count(max_val+1, 0);
    for(int num: arr) count[num]++;
    vector<int> sorted;
    for(int i=0; i<=max_val; i++){
        for(int j=0; j<count[i]; j++)
            sorted.push_back(i);
    }
    return sorted;
}

void printArray(const vector<int> &arr){
    for(int num: arr) cout << num << " ";
    cout << endl;
}

int main(){
    vector<int> arr = {4, 2, 2, 8, 3, 3, 1};
    cout << "Original: ";
    printArray(arr);

    vector<int> sorted = countingSort(arr);
    cout << "Counting Sorted: ";
    printArray(sorted);

    return 0;
}

File: java/sorting/CountingSort.java
import java.util.Arrays;

public class CountingSort {
    public static int[] countingSort(int[] arr) {
        if(arr.length == 0) return arr;
        int max = Arrays.stream(arr).max().getAsInt();
        int[] count = new int[max+1];
        for(int num : arr) count[num]++;
        int index = 0;
        for(int i=0; i<=max; i++) {
            while(count[i] > 0) {
                arr[index++] = i;
                count[i]--;
            }
        }
        return arr;
    }

    public static void main(String[] args) {
        int[] arr = {4, 2, 2, 8, 3, 3, 1};
        System.out.println("Original: " + Arrays.toString(arr));
        countingSort(arr);
        System.out.println("Counting Sorted: " + Arrays.toString(arr));
    }
}

